### PR TITLE
Ruby: Add Server-Side Request Forgery query

### PR DESCRIPTION
**Pull request by hmac**
The query finds cases where user input flows to the URL in an outgoing HTTP request.

We use the same sanitizer as with URL redirection: prefixing user input with some non-user-controlled string is treated as sanitizing it, because the user no longer has complete control over the URL of the request.

This PR also fixes a bug with `CallExprCfgNode.getKeywordArgument`, which was previously not giving any results.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/security/ServerSideRequestForgeryQuery.qll [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/**
+ * Provides a taint-tracking configuration for detecting
+ * "Server side request forgery" vulnerabilities.
+ *
+ * Note, for performance reasons: only import this file if `Configuration` is needed,
+ * otherwise `ServerSideRequestForgeryCustomizations` should be imported instead.
+ */
+
+import codeql.ruby.DataFlow::DataFlow::PathGraph
```
Comment (intrigus-lgtm) [inline ruby/ql/lib/codeql/ruby/security/ServerSideRequestForgeryQuery.qll:9]: Is this missing `import ruby` as the first statement?
AFAIK `import ruby` should be the first statement for [caching](https://github.com/github/codeql/commit/deb6a6e5c69ee7ea54b97561554ed4c0cee471fe) (?) reasons.

Comment (hmac) [reply ruby/ql/lib/codeql/ruby/security/ServerSideRequestForgeryQuery.qll:9]: Ah interesting! I didn't know that, thanks for pointing it out!

Comment (hmac) [reply ruby/ql/lib/codeql/ruby/security/ServerSideRequestForgeryQuery.qll:9]: Actually, I wonder if that only applies to query files. The corresponding `ServerSideRequestForgery.ql` file does `import ruby`, so I think we may be OK.

Comment (intrigus-lgtm) [reply ruby/ql/lib/codeql/ruby/security/ServerSideRequestForgeryQuery.qll:9]: I thought I've read this somewhere, but I'm not sure anymore.
If you should find a definite answer, please share it with me :)

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/security/ServerSideRequestForgeryQuery.qll:9]: @hvitved ^

**Review by alexrford — COMMENTED** [review]
Looks good - I've left some suggestions for extending coverage for some of the HTTP client libraries here which are a bit broader than just the new query here, so I don't mind if these are pushed into a separate issue/PR.

There seem to be some performance regressions in the DCA run that would be worth investigating.

File: ruby/ql/lib/codeql/ruby/frameworks/http_clients/Httparty.qll [hunk 2]
```diff
@@ -28,6 +28,8 @@ class HttpartyRequest extends HTTP::Client::Request::Range {
     this = requestUse.asExpr().getExpr()
   }
 
+  override DataFlow::Node getURL() { result = requestUse.getArgument(0) }
```
[the remark this answers is not in the document]
Comment (hmac) [reply ruby/ql/lib/codeql/ruby/frameworks/http_clients/Httparty.qll:31]: Going from those examples, this seems to be used when `include`ing `HTTParty`, which is not something I've modelled at all yet - I've updated the comment on this class to include a `base_uri` call in the TODO section, so we remember to model it if/when we come back to flesh this library out further. Does that sound OK to you?

[the remark this answers is not in the document]
Comment (alexrford) [reply ruby/ql/lib/codeql/ruby/frameworks/http_clients/Httparty.qll:32]: Yep, sounds good to me.

**Pull request merged**